### PR TITLE
Port for newer LLVM

### PR DESCRIPTION
**Pull request by mchalupa**
This PR just ports the LLVM CL to newer LLVM, it should not introduce any functional changes.

### CODE REVIEW

File: cl/llvm/clplug.cc [hunk 1]
```diff
@@ -32,14 +32,27 @@ extern "C" {
...
+#   define LLVM_HOST_4_OR_NEWER
+#endif
+
+
+
 #ifdef LLVM_HOST_3_7_OR_NEWER
 #   include "llvm/Transforms/Scalar.h" // createLowerSwitchPass
 #   include "llvm/IR/LegacyPassManager.h"
 #   include "llvm/Transforms/IPO/PassManagerBuilder.h"
+#   include "llvm/Transforms/Utils.h"
```
Comment (kdudka) [inline cl/llvm/clplug.cc]: This header file for some reason does not exist on any of my installations of llvm, tried with 4.0.1, 5.0.2, and 6.0.1:
```
cl/llvm/clplug.cc:55:13: fatal error: llvm/Transforms/Utils.h: No such file or directory
 #   include "llvm/Transforms/Utils.h"
             ^~~~~~~~~~~~~~~~~~~~~~~~~
```

Comment (kdudka) [reply cl/llvm/clplug.cc]: I forgot to say that it compiles successfully with llvm-5.0.2 and llvm-6.0.1 if I just comment out the mentioned include directive.

Comment (mchalupa) [reply cl/llvm/clplug.cc]: Sorry, my bad. I did not check the version of LLVM correctly. Should be fixed now.

**Comment by versokova** [comment]
It looks good. But it has an interesting behavior. If Predator starts and the sl pass has detected some errors, run is OK. If the program is error-free, the run ends on segmentation fault (after the analysis has end, last note is '.*note: clEasyRun() took 0.001 s.*' and segfault). Tested on clang 7.0.1 and clang 8.0.0.

**Comment by kdudka** [comment]
Maybe a stupid question but how can I compile the `passes` subproject against llvm-8.0.0?

The build fails in my environment with a lot of compile-time errors:
```
make[2]: Entering directory '/home/kdudka/predator/passes-src/passes_build'
[  9%] Building CXX object CMakeFiles/passes_test.dir/test.cc.o
/usr/bin/c++  -Dpasses_test_EXPORTS  -fPIC   -std=c++11 -fPIC -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -o CMakeFiles/passes_test.dir/test.cc.o -c /home/kdudka/predator/passes-src/passes/test.cc
[ 18%] Building CXX object CMakeFiles/passes_test.dir/global_vars.cc.o
/usr/bin/c++  -Dpasses_test_EXPORTS  -fPIC   -std=c++11 -fPIC -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -o CMakeFiles/passes_test.dir/global_vars.cc.o -c /home/kdudka/predator/passes-src/passes/global_vars.cc
/home/kdudka/predator/passes-src/passes/global_vars.cc: In member function ‘virtual bool GlobalVarsPass::runOnModule(llvm::Module&)’:
/home/kdudka/predator/passes-src/passes/global_vars.cc:135:44: error: ‘getGlobalContext’ was not declared in this scope
  135 |  BasicBlock* blockGVF = BasicBlock::Create(getGlobalContext(), "entry", initGVF);
      |                                            ^~~~~~~~~~~~~~~~
make[2]: *** [CMakeFiles/passes_test.dir/build.make:76: CMakeFiles/passes_test.dir/global_vars.cc.o] Error 1
[ 27%] Building CXX object CMakeFiles/passes_test.dir/lower_mem.cc.o
/usr/bin/c++  -Dpasses_test_EXPORTS  -fPIC   -std=c++11 -fPIC -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -o CMakeFiles/passes_test.dir/lower_mem.cc.o -c /home/kdudka/predator/passes-src/passes/lower_mem.cc
/home/kdudka/predator/passes-src/passes/lower_mem.cc: In member function ‘bool LowerMemIntrinsicPass::testMem(llvm::MemIntrinsic*)’:
/home/kdudka/predator/passes-src/passes/lower_mem.cc:63:23: error: ‘class llvm::MemIntrinsic’ has no member named ‘getAlignment’; did you mean ‘getRetAlignment’?
   63 |  unsigned align = in->getAlignment();
      |                       ^~~~~~~~~~~~
      |                       getRetAlignment
/home/kdudka/predator/passes-src/passes/lower_mem.cc: In member function ‘bool LowerMemIntrinsicPass::testMemTransfer(llvm::MemTransferInst*)’:
/home/kdudka/predator/passes-src/passes/lower_mem.cc:90:23: error: ‘class llvm::MemTransferInst’ has no member named ‘getAlignment’; did you mean ‘getRetAlignment’?
   90 |  unsigned align = in->getAlignment();
      |                       ^~~~~~~~~~~~
      |                       getRetAlignment
/home/kdudka/predator/passes-src/passes/lower_mem.cc: In member function ‘bool LowerMemIntrinsicPass::replaceMemTransfer(llvm::MemTransferInst*)’:
/home/kdudka/predator/passes-src/passes/lower_mem.cc:126:23: error: ‘class llvm::MemTransferInst’ has no member named ‘getAlignment’; did you mean ‘getRetAlignment’?
  126 |  unsigned align = in->getAlignment();
      |                       ^~~~~~~~~~~~
      |                       getRetAlignment
/home/kdudka/predator/passes-src/passes/lower_mem.cc:129:42: error: ‘class llvm::GlobalVariable’ has no member named ‘hasUnnamedAddr’; did you mean ‘getUnnamedAddr’?
  129 |  if (gv && gv->isConstant()==true && gv->hasUnnamedAddr()==true && gv->hasPrivateLinkage()) {
      |                                          ^~~~~~~~~~~~~~
      |                                          getUnnamedAddr
/home/kdudka/predator/passes-src/passes/lower_mem.cc: In member function ‘void LowerMemIntrinsicPass::printMemCpyInst(llvm::MemCpyInst*)’:
/home/kdudka/predator/passes-src/passes/lower_mem.cc:219:23: error: ‘class llvm::MemCpyInst’ has no member named ‘getAlignment’; did you mean ‘getRetAlignment’?
  219 |  unsigned align = in->getAlignment();
      |                       ^~~~~~~~~~~~
      |                       getRetAlignment
/home/kdudka/predator/passes-src/passes/lower_mem.cc:236:29: error: ‘class llvm::MemCpyInst’ has no member named ‘getAlignment’; did you mean ‘getRetAlignment’?
  236 |  errs() << "align: " << in->getAlignment() << " / ";
      |                             ^~~~~~~~~~~~
      |                             getRetAlignment
make[2]: *** [CMakeFiles/passes_test.dir/build.make:89: CMakeFiles/passes_test.dir/lower_mem.cc.o] Error 1
[ 36%] Building CXX object CMakeFiles/passes_test.dir/lower_select.cc.o
/usr/bin/c++  -Dpasses_test_EXPORTS  -fPIC   -std=c++11 -fPIC -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -o CMakeFiles/passes_test.dir/lower_select.cc.o -c /home/kdudka/predator/passes-src/passes/lower_select.cc
/home/kdudka/predator/passes-src/passes/lower_select.cc: In member function ‘void LowerSelectPass::replaceSelect(llvm::SelectInst*)’:
/home/kdudka/predator/passes-src/passes/lower_select.cc:57:48: error: invalid conversion from ‘const char*’ to ‘unsigned int’ [-fpermissive]
   57 |  Value *newPtr = new AllocaInst(in->getType(), "", /*before*/bbThis->getFirstNonPHI());
      |                                                ^~
      |                                                |
      |                                                const char*
In file included from /home/kdudka/predator/passes-src/passes/lower_select.hh:12,
                 from /home/kdudka/predator/passes-src/passes/lower_select.cc:9:
/usr/include/llvm/IR/Instructions.h:70:42: note:   initializing argument 2 of ‘llvm::AllocaInst::AllocaInst(llvm::Type*, unsigned int, llvm::Value*, const llvm::Twine&, llvm::Instruction*)’
   70 |   explicit AllocaInst(Type *Ty, unsigned AddrSpace,
      |                                 ~~~~~~~~~^~~~~~~~~
make[2]: *** [CMakeFiles/passes_test.dir/build.make:102: CMakeFiles/passes_test.dir/lower_select.cc.o] Error 1
make[2]: Target 'CMakeFiles/passes_test.dir/build' not remade because of errors.
make[2]: Leaving directory '/home/kdudka/predator/passes-src/passes_build'
make[1]: *** [CMakeFiles/Makefile2:105: CMakeFiles/passes_test.dir/all] Error 2
```

**Comment by versokova** [comment]
The same thing for llvm-7.0.1. The Predator only needs global-vars, it compiles - but it doesn't work correctly. I deleted it from dependences.

**Comment by mchalupa** [comment]
Yes, it is because of this commit in LLVM: a92bcbb2c8c55 (svn rev 328315). `lower_mem.cc` uses the removed methods. Actually, LLVM has some LowerMem pass, maybe it could be better to use the upstream one if it does the same thing (I do not know if it does exactly what this one).

**Comment by mchalupa** [comment]
@versokova: I cannot reproduce the segfault. It looks like you have a fix, so could you just create a follow-up patch that fixes it? I do not want to be blind-coding.

**Comment by versokova** [comment]
I suppose it depends on Darwin. Thanks for trying. PR is OK.

**Comment by kdudka** [comment]
OK, I did not get it fully working but it compiles fine and none of the observed problems seems to be triggered by changes from this pull request.  Let's merge it as it is.

@mchalupa I should have said in the beginning that the LLVM support in Predator is experimental.  You need to use the GCC front-end to get reliable results out of Predator.  Anyway, thank you for contributing the fixes!

**Pull request closed without merging**